### PR TITLE
Infer inequality predicates in joins

### PR DESCRIPTION
**Pull request by ClarenceThreepwood**
Presto has a good mechanism for inferring equality predicates through transitive closure from join predicates. These inferred predicates may be pushed down to the table scan leading to improved performance. Among our customers, we have seen a few cases where automatically inferring inequality predicates and pushing them to the table scan would lead to significant improvement. This PR is a first cut at attempting to infer inequality predicates.

Controlled by a feature flag "infer_inequality_predicates"

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+                referencedVariables = VariablesExtractor.extractUnique(secondConjunct.getArguments().get(0));
+            }
+
+            if (referencedVariables.isEmpty()) {
+                inferredFirstArgument = Optional.of(secondConjunct.getArguments().get(0));
+                inferredSecondArgument = Optional.of(firstConjunct.getArguments().get(1));
+            }
+        }
+
+        if (!inferredFirstArgument.isPresent() || (outerVariables.isPresent() && Iterables.any(variablesReferencedInInferredPredicate.build(), in(outerVariables.get())))) {
```
Comment (aaneja) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java:150]: For safety also add `!inferredSecondArgument.isPresent()` to the OR list

**Review by aaneja — CHANGES_REQUESTED** [review]
LGTM except for a few comments

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        this.expressionEquivalence = expressionEquivalence;
+        this.outerVariables = outerVariables;
+    }
+
+    // Basic idea is to pairwise compare each predicate with all other predicates. If the join is an outer join, we need to make sure
+    // any inferred predicate is on the inner side only (i.e. by using 'outerVariables').
+    public ImmutableSet<RowExpression> inferInequalities()
+    {
+        inequalityExpressions.forEach(this::validateConjunct);
+        IntStream.range(0, inequalityExpressions.size() - 1).forEach(this::compareInequalityAtIndexWithRemaining);
```
[the remark this answers is not in the document]
Comment (ClarenceThreepwood) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java]: I'm not particular either way. Changed it to the more traditional for-loop

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+        Optional<OperatorType> inferenceOperatorType = getComparisonInferenceOperatorType(secondOperatorType, firstOperatorType);
+        if (!inferenceOperatorType.isPresent()) {
+            return;
+        }
+
+        Optional<RowExpression> inferredFirstArgument = Optional.empty();
+        Optional<RowExpression> inferredSecondArgument = Optional.empty();
+        ImmutableSet.Builder<VariableReferenceExpression> variablesReferencedInInferredPredicate = ImmutableSet.builder();
+
+        // For two inequalities "a < b" and "b < 5" we want to infer "a < 5"
```
[the remark this answers is not in the document]
Comment (ClarenceThreepwood) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java:113]: It does not and I will defer that for later since this first cut of inequality inferencing does not consider variable scope

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExpressionEquivalence.java [hunk 4]
```diff
@@ -146,7 +146,7 @@ public RowExpression visitCall(CallExpression call, Void context)
 
             QualifiedObjectName callName = functionAndTypeManager.getFunctionMetadata(call.getFunctionHandle()).getName();
 
-            if (callName.equals(EQUAL.getFunctionName()) || callName.equals(NOT_EQUAL.getFunctionName()) || callName.equals(IS_DISTINCT_FROM.getFunctionName())) {
+            if (callName.equals(EQUAL.getFunctionName()) || callName.equals(NOT_EQUAL.getFunctionName()) || callName.equals(IS_DISTINCT_FROM.getFunctionName()) || callName.equals(ADD.getFunctionName()) || callName.equals(MULTIPLY.getFunctionName())) {
```
[the remark this answers is not in the document]
Comment (ClarenceThreepwood) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExpressionEquivalence.java:151]: Good catch - thanks!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (aaneja) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java:127]: Can you add a test case in `TestInequalityInference#testChainOfTransitivity` that calls this out ?

**Review by aaneja — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ClarenceThreepwood) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java:127]: It is already there - look at TestInequalityInference:147

**Comment by yzhang1991** [comment]
Hi @feilong-liu @rschlussel, could you help take a look at this PR? Thanks so much!!

**Comment by rschlussel** [comment]
> @rschlussel - any thoughts on this?

Reviewing now

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+        // Make sure one input has no column/variable references since we only want to infer predicates with constants
+        // LHS : Left hand side
+        // RHS : Right hand side
+        RowExpression firstConjunctLHS = firstConjunct.getArguments().get(0);
+        RowExpression firstConjunctRHS = firstConjunct.getArguments().get(1);
+        RowExpression secondConjunctLHS = secondConjunct.getArguments().get(0);
+        RowExpression secondConjunctRHS = secondConjunct.getArguments().get(1);
+        if (expressionEquivalence.areExpressionsEquivalent(firstConjunctRHS, secondConjunctLHS)) {
+            Set<VariableReferenceExpression> referencedVariables = VariablesExtractor.extractUnique(firstConjunctLHS);
+            if (referencedVariables.isEmpty()) {
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java:149]: This is a lot of repeated logic for each case and a bit hard to follow.  I would extract a method for getting the variablesReferencedInInferredPredicate.

e.g. something like 
```
if (expressionEquivalence.areExpressionsEquivalent(firstConjunctRHS, secondConjunctLHS)) {
     variablesReferencedInInferredPredicate = getVariablesReferencedInInferredPredicate(firstConjunctLHS, secondConjunctRHS);
      if(!variablesReferencedInInferredPredicate.isEmpty()) {
           inferredFirstArgument = Optional.of(firstConjunctLHS);
           inferredSecondArgument = Optional.of(secondConjunctRHS);
       }
}
else if(expressionEquivalence.areExpressionsEquivalent(firstConjunctRHS, secondConjunctLHS)) {
     variablesReferencedInInferredPredicate = getVariablesReferencedInInferredPredicate(firstConjunctRHS, secondConjunctLHS);
      if(!variablesReferencedInInferredPredicate.isEmpty()) {
           inferredFirstArgument = Optional.of(firstConjunctRHS);
           inferredSecondArgument = Optional.of(secondConjunctLHS);
       }
}
```
 and then
```
private static Set<VariableReferenceExpression> getVariablesReferencedInInferredPredicate(RowExpression firstConjunct, RowExpression secondConjunct)
{
    Set<VariableReferenceExpression> firstConjunctReferencedVariables = VariablesExtractor.extractUnique(firstConjunct);
     if(firstConjunctReferencedVariables.isEmpty()) {
         return VariablesExtractor.extractUnique(secondConjunct);
     } 
    Set<VariableReferenceExpression> secondConjunctReferencedVariables 
= VariablesExtractor.extractUnique(secondConjunct);
    if(secondConjunctReferencedVariables.isEmpty()) 
    { 
       return firstConjunctReferencedVariables;
    }
   // the inferred predicate cannot reference variables from both expressions
  // return empty to indicate that there is no valid inferred predicate
   return ImmutableSet.of();
 }
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    }
+
+    // Basic idea is to pairwise compare each predicate with all other predicates.
+    // Since any inferred predicates may trigger other inferences, keep comparing until no new predicates are generated
+    public ImmutableSet<RowExpression> inferInequalities()
+    {
+        if (inequalityExpressions.size() < 2) {
+            return ImmutableSet.of();
+        }
+        inequalityExpressions.forEach(this::validateConjunct);
```
[the remark this answers is not in the document]
Comment (ClarenceThreepwood) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java]: I was mildy concerned about the new predicates that we generate and wanted to ensure they conform to this pattern. But this should be safe enough

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+        Set<RowExpression> inferredInequalityExpressions = new HashSet<>();
+        Set<RowExpression> inequalitiesInferredInCurrentTraversal = new HashSet<>();
+        do {
```
[the remark this answers is not in the document]
Comment (ClarenceThreepwood) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java]: Makes sense to partition the predicates into new combinations and ones we have already processed. Thanks!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java (hunk 5 above)
Comment (ClarenceThreepwood) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java]: Thanks! This is easier to parse

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+            }
+        }
+
+        if (!inferredFirstArgument.isPresent() ||
+                !inferredSecondArgument.isPresent() ||
+                (outerVariables.isPresent() && Iterables.any(variablesReferencedInInferredPredicate, in(outerVariables.get())))) {
+            return Optional.empty();
+        }
+
+        // Build the new predicate and add to 'inferredInequalityExpressions'
```
[the remark this answers is not in the document]
Comment (ClarenceThreepwood) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java]: From an earlier iteration. Cleaned up now

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+
+        // Get the inference operator, or empty if no inference possible.
+        Optional<OperatorType> inferenceOperatorType = getComparisonInferenceOperatorType(secondOperatorType, firstOperatorType);
+        if (!inferenceOperatorType.isPresent()) {
+            return Optional.empty();
+        }
+
+        Optional<RowExpression> inferredFirstArgument = Optional.empty();
+        Optional<RowExpression> inferredSecondArgument = Optional.empty();
+        Set<VariableReferenceExpression> variablesReferencedInInferredPredicate = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (ClarenceThreepwood) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java]: That would lead to a "variable not initialized" error when we use it later. But with the suggested rewrite, I think the code is easier to parse, and its usage is guarded by "inferredFirstArgument.isPresent()". Wdyt?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+    public InequalityInference(Set<RowExpression> inequalityExpressions, FunctionAndTypeManager functionAndTypeManager, ExpressionEquivalence expressionEquivalence, Optional<Collection<VariableReferenceExpression>> outerVariables)
+    {
+        if (inequalityExpressions.stream()
+                .anyMatch(e -> !isOperation(e, LESS_THAN, functionAndTypeManager) && !isOperation(e, LESS_THAN_OR_EQUAL, functionAndTypeManager))) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "all inequality expressions for inference must be < or <=");
+        }
+        this.inequalityExpressions = requireNonNull(inequalityExpressions, "inequalityExpressions is null");
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+        this.expressionEquivalence = requireNonNull(expressionEquivalence, "expressionEquivalence is null");
+        this.outerVariables = requireNonNull(outerVariables, "outerVariables in null");
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java:76]: typo: in -> is

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+
+        Set<RowExpression> allInferredInequalities = new HashSet<>();
+        Set<RowExpression> inequalitiesInferredInCurrentTraversal = new HashSet<>(inequalityExpressions);
+        Set<Set<RowExpression>> exploredCombinations = new HashSet();
+
+        while (!allInferredInequalities.containsAll(inequalitiesInferredInCurrentTraversal)) {
+            allInferredInequalities.addAll(inequalitiesInferredInCurrentTraversal);
+            Set<Set<RowExpression>> newCombinations = Sets.combinations(allInferredInequalities, 2).stream()
+                    .filter(subset -> !exploredCombinations.contains(subset))
+                    .collect(Collectors.toSet());
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java:95]: nit: toImmutableSet()

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+            }
+
+            inequalitiesInferredInCurrentTraversal = newCombinations.stream()
+                    .map(pair -> {
+                        Iterator<RowExpression> it = pair.iterator();
+                        return compareAndExtractInequalities(it.next(), it.next());
+                    })
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .collect(Collectors.toSet());
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java:109]: nit: toImmutableSet()

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+
+        while (!allInferredInequalities.containsAll(inequalitiesInferredInCurrentTraversal)) {
+            allInferredInequalities.addAll(inequalitiesInferredInCurrentTraversal);
+            Set<Set<RowExpression>> newCombinations = Sets.combinations(allInferredInequalities, 2).stream()
+                    .filter(subset -> !exploredCombinations.contains(subset))
+                    .collect(Collectors.toSet());
+
+            for (Set<RowExpression> pair : newCombinations) {
+                Iterator<RowExpression> it = pair.iterator();
+                compareAndExtractInequalities(it.next(), it.next()).ifPresent(inequalitiesInferredInCurrentTraversal::add);
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java:99]: we don't want to add to the existing inequalitiesInferredInCurrentTraversal.  We want to replace it with a new set.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+            }
+
+            inequalitiesInferredInCurrentTraversal = newCombinations.stream()
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java:102]: I think the code starting from line 96 and the code starting from here are trying to do the same thing, and we need only one or the other.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+        // LHS : Left hand side
+        // RHS : Right hand side
+        RowExpression firstConjunctLHS = firstConjunct.getArguments().get(0);
+        RowExpression firstConjunctRHS = firstConjunct.getArguments().get(1);
+        RowExpression secondConjunctLHS = secondConjunct.getArguments().get(0);
+        RowExpression secondConjunctRHS = secondConjunct.getArguments().get(1);
+
+        Optional<RowExpression> inferredFirstArgument = Optional.empty();
+        Optional<RowExpression> inferredSecondArgument = Optional.empty();
+        Set<VariableReferenceExpression> variablesReferencedInInferredPredicate = new HashSet<>();
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/InequalityInference.java:147]: set this to ImmutableSet.of() since we never want to add anything to it.

**Review by rschlussel — APPROVED** [review]
